test snapshot fast-forward attack recovery

kairoaraujo opened the pull request:
This test simulates the snapshot fast-forward attack recovery.
It simulates that the snapshot keys were compromised, the attacker
generated a new high version of the snapshot.

The repository generates new keys for snapshot and timestamp and
rollbacks the snapshot version to the initial version.

Signed-off-by: Kairo de Araujo <kdearaujo@vmware.com>

Part of #1713 

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
test snapshot fast-forward attack recovery

This test simulates the snapshot fast-forward attack recovery.
It simulates that the snapshot keys were compromised, the attacker
generated a new high version of the snapshot.

The repository generates new keys for snapshot and timestamp and
rollbacks the snapshot version to the initial version.

Signed-off-by: Kairo de Araujo <kdearaujo@vmware.com>

jku commented on tests/test_updater_top_level_update.py:
```diff
@@ -424,6 +424,47 @@ def test_new_snapshot_version_rollback(self) -> None:
+
+        # client refreshes the metadata and see the new snapshot version
+        self._run_refresh()
+        self._assert_version_equals(Snapshot.type, 99999)
+
+        # repo add new snapshot and timestamp keys and recovers snapshot version
+        self.sim.root.roles["snapshot"].keyids.clear()
```
I didn't even notice this in the previous PR but it's a bit inconsistent that both `Snapshot.type` and `"snapshot"` are used... could just use Snapshot.type everywhere, or does it look worse like that?

kairoaraujo replied:
Any objection about fixing it for the previous PR on this PR?

jku replied:
if it looks fine to you, go ahead

jku commented on tests/test_updater_top_level_update.py:
```diff
@@ -424,6 +424,47 @@ def test_new_snapshot_version_rollback(self) -> None:
+        timestamp_key, timestamp_signer = self.sim.create_key()
+        self.sim.root.add_key("timestamp", timestamp_key)
+        self.sim.add_signer("timestamp", timestamp_signer)
+        self.sim.snapshot.version = 1
+        self.sim.root.version += 1
+        self.sim.publish_root()
+
+        self.sim.update_timestamp()
```
This looks correct but for clarity following might be better (root update just contains the key changes: then we rollback snapshot version, signing with the new keys)

```suggestion
        self.sim.root.version += 1
        self.sim.publish_root()

        self.sim.snapshot.version = 1
        self.sim.update_timestamp()
```

jku reviewed (commented):
Yep, agreed on the implementation and listed requirements. Left some questions / comments

Commit:
remove roles names as str, snapshot order

This commit removes the role names as strings. Also do a slight
change for clarity.

Signed-off-by: Kairo de Araujo <kdearaujo@vmware.com>

jku reviewed (approved):
Looks great, thanks.
